File: examples/flutter/lib/services/authon_service.dart
import 'dart:convert';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:url_launcher/url_launcher.dart';

const String _baseUrl = 'https://api.authon.dev';
const String _storageKeyToken = 'authon_access_token';
const String _storageKeyRefresh = 'authon_refresh_token';

class AuthonUser {
  final String id;
  final String? email;
  final bool emailVerified;
  final String? phone;
  final bool phoneVerified;
  final String? username;
  final String? firstName;
  final String? lastName;
  final String? displayName;
  final String? avatarUrl;
  final bool banned;
  final List<ExternalAccount> externalAccounts;
  final DateTime createdAt;
  final DateTime updatedAt;

  const AuthonUser({
    required this.id,
    this.email,
    this.emailVerified = false,
    this.phone,
    this.phoneVerified = false,
    this.username,
    this.firstName,
    this.lastName,
    this.displayName,
    this.avatarUrl,
    this.banned = false,
    this.externalAccounts = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory AuthonUser.fromJson(Map<String, dynamic> json) {
    return AuthonUser(
      id: json['id'] as String,
      email: json['email'] as String?,
      emailVerified: json['emailVerified'] as bool? ?? false,
      phone: json['phone'] as String?,
      phoneVerified: json['phoneVerified'] as bool? ?? false,
      username: json['username'] as String?,
      firstName: json['firstName'] as String?,
      lastName: json['lastName'] as String?,
      displayName: json['displayName'] as String?,
      avatarUrl: json['avatarUrl'] as String?,
      banned: json['banned'] as bool? ?? false,
      externalAccounts: (json['externalAccounts'] as List<dynamic>?)
              ?.map((e) => ExternalAccount.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  String get initials {
    if (displayName != null && displayName!.isNotEmpty) {
      final parts = displayName!.trim().split(' ');
      if (parts.length >= 2) {
        return '${parts.first[0]}${parts.last[0]}'.toUpperCase();
      }
      return displayName![0].toUpperCase();
    }
    if (email != null && email!.isNotEmpty) {
      return email![0].toUpperCase();
    }
    return '?';
  }

  String get displayLabel => displayName ?? email?.split('@').first ?? username ?? id;
}

class ExternalAccount {
  final String provider;
  final String providerId;
  final String? email;

  const ExternalAccount({
    required this.provider,
    required this.providerId,
    this.email,
  });

  factory ExternalAccount.fromJson(Map<String, dynamic> json) {
    return ExternalAccount(
      provider: json['provider'] as String,
      providerId: json['providerId'] as String,
      email: json['email'] as String?,
    );
  }
}

class AuthonSession {
  final String id;
  final String userId;
  final String status;
  final String? userAgent;
  final String? ipAddress;
  final DateTime lastActiveAt;
  final DateTime expireAt;
  final DateTime createdAt;

  const AuthonSession({
    required this.id,
    required this.userId,
    required this.status,
    this.userAgent,
    this.ipAddress,
    required this.lastActiveAt,
    required this.expireAt,
    required this.createdAt,
  });

  factory AuthonSession.fromJson(Map<String, dynamic> json) {
    return AuthonSession(
      id: json['id'] as String,
      userId: json['userId'] as String,
      status: json['status'] as String,
      userAgent: json['userAgent'] as String?,
      ipAddress: json['ipAddress'] as String?,
      lastActiveAt: DateTime.parse(json['lastActiveAt'] as String),
      expireAt: DateTime.parse(json['expireAt'] as String),
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }
}

class MfaStatus {
  final bool enabled;
  final int backupCodesRemaining;

  const MfaStatus({
    required this.enabled,
    required this.backupCodesRemaining,
  });

  factory MfaStatus.fromJson(Map<String, dynamic> json) {
    return MfaStatus(
      enabled: json['enabled'] as bool? ?? false,
      backupCodesRemaining: json['backupCodesRemaining'] as int? ?? 0,
    );
  }
}

class MfaSetupData {
  final String secret;
  final String qrCodeUrl;
  final List<String> backupCodes;

  const MfaSetupData({
    required this.secret,
    required this.qrCodeUrl,
    required this.backupCodes,
  });

  factory MfaSetupData.fromJson(Map<String, dynamic> json) {
    return MfaSetupData(
      secret: json['secret'] as String,
      qrCodeUrl: json['qrCodeUrl'] as String? ?? '',
      backupCodes: (json['backupCodes'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
    );
  }
}

class AuthonException implements Exception {
  final int statusCode;
  final String message;
  final String? code;

  const AuthonException({
    required this.statusCode,
    required this.message,
    this.code,
  });

  @override
  String toString() => code != null ? '$code: $message' : message;
}

class AuthonMfaRequiredException implements Exception {
  final String mfaToken;
  const AuthonMfaRequiredException(this.mfaToken);
}

class AuthonService {
  final String publishableKey;
  final FlutterSecureStorage _storage;
  final http.Client _http;

  String? _accessToken;
  String? _refreshToken;
  AuthonUser? _currentUser;

  AuthonService({
    required this.publishableKey,
    FlutterSecureStorage? storage,
    http.Client? httpClient,
  })  : _storage = storage ?? const FlutterSecureStorage(),
        _http = httpClient ?? http.Client();

  AuthonUser? get currentUser => _currentUser;
  bool get isSignedIn => _currentUser != null && _accessToken != null;

  Future<void> initialize() async {
    _accessToken = await _storage.read(key: _storageKeyToken);
    _refreshToken = await _storage.read(key: _storageKeyRefresh);
    if (_accessToken != null) {
      try {
        await _loadCurrentUser();
      } catch (_) {
        await _clearTokens();
      }
    }
  }

  Future<AuthonUser> signInWithEmail(String email, String password) async {
    final body = await _request(
      'POST',
      '/v1/auth/sign-in',
      body: {'email': email, 'password': password},
      requiresAuth: false,
    );

    if (body['mfaRequired'] == true) {
      throw AuthonMfaRequiredException(body['mfaToken'] as String);
    }

    await _storeTokens(
      accessToken: body['accessToken'] as String,
      refreshToken: body['refreshToken'] as String?,
    );
    _currentUser = AuthonUser.fromJson(body['user'] as Map<String, dynamic>);
    return _currentUser!;
  }

  Future<AuthonUser> signUpWithEmail(
    String email,
    String password, {
    String? displayName,
  }) async {
    final body = await _request(
      'POST',
      '/v1/auth/sign-up',
      body: {
        'email': email,
        'password': password,
        if (displayName != null) 'displayName': displayName,
      },
      requiresAuth: false,
    );

    await _storeTokens(
      accessToken: body['accessToken'] as String,
      refreshToken: body['refreshToken'] as String?,
    );
    _currentUser = AuthonUser.fromJson(body['user'] as Map<String, dynamic>);
    return _currentUser!;
  }

  Future<AuthonUser> verifyMfa(String mfaToken, String code) async {
    final body = await _request(
      'POST',
      '/v1/auth/mfa/verify',
      body: {'mfaToken': mfaToken, 'code': code},
      requiresAuth: false,
    );

    await _storeTokens(
      accessToken: body['accessToken'] as String,
      refreshToken: body['refreshToken'] as String?,
    );
    _currentUser = AuthonUser.fromJson(body['user'] as Map<String, dynamic>);
    return _currentUser!;
  }

  Future<void> signOut() async {
    try {
      await _request('POST', '/v1/auth/sign-out');
    } catch (_) {}
    await _clearTokens();
    _currentUser = null;
  }

  Future<void> sendPasswordResetEmail(String email) async {
    await _request(
      'POST',
      '/v1/auth/forgot-password',
      body: {'email': email},
      requiresAuth: false,
    );
  }

  Future<AuthonUser> updateProfile({
    String? displayName,
    String? avatarUrl,
    String? phone,
    String? firstName,
    String? lastName,
  }) async {
    final body = await _request(
      'PATCH',
      '/v1/auth/me',
      body: {
        if (displayName != null) 'displayName': displayName,
        if (avatarUrl != null) 'avatarUrl': avatarUrl,
        if (phone != null) 'phone': phone,
        if (firstName != null) 'firstName': firstName,
        if (lastName != null) 'lastName': lastName,
      },
    );
    _currentUser = AuthonUser.fromJson(body);
    return _currentUser!;
  }

  Future<void> deleteAccount() async {
    await _request('DELETE', '/v1/auth/me');
    await _clearTokens();
    _currentUser = null;
  }

  Future<MfaStatus> getMfaStatus() async {
    final body = await _request('GET', '/v1/auth/mfa/status');
    return MfaStatus.fromJson(body);
  }

  Future<MfaSetupData> setupMfa() async {
    final body = await _request('POST', '/v1/auth/mfa/setup');
    return MfaSetupData.fromJson(body);
  }

  Future<void> verifyMfaSetup(String code) async {
    await _request(
      'POST',
      '/v1/auth/mfa/setup/verify',
      body: {'code': code},
    );
  }

  Future<void> disableMfa(String code) async {
    await _request(
      'POST',
      '/v1/auth/mfa/disable',
      body: {'code': code},
    );
  }

  Future<List<String>> regenerateBackupCodes(String code) async {
    final body = await _request(
      'POST',
      '/v1/auth/mfa/backup-codes/regenerate',
      body: {'code': code},
    );
    return (body['backupCodes'] as List<dynamic>)
        .map((e) => e as String)
        .toList();
  }

  Future<List<AuthonSession>> listSessions() async {
    final body = await _request('GET', '/v1/auth/sessions');
    final data = body['data'] as List<dynamic>? ?? body['sessions'] as List<dynamic>? ?? [];
    return data
        .map((e) => AuthonSession.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<void> revokeSession(String sessionId) async {
    await _request('DELETE', '/v1/auth/sessions/$sessionId');
  }

  Future<void> launchOAuthProvider(String provider) async {
    final redirectUri = Uri.encodeComponent('authon-flutter-example://oauth');
    final url = Uri.parse(
      '$_baseUrl/v1/auth/oauth/$provider?publishableKey=$publishableKey&redirectUri=$redirectUri',
    );
    if (!await launchUrl(url, mode: LaunchMode.externalApplication)) {
      throw AuthonException(
        statusCode: 0,
        message: 'Could not launch OAuth for $provider',
      );
    }
  }

  Future<AuthonUser> handleOAuthCallback(String code, String state) async {
    final body = await _request(
      'POST',
      '/v1/auth/oauth/callback',
      body: {'code': code, 'state': state},
      requiresAuth: false,
    );

    await _storeTokens(
      accessToken: body['accessToken'] as String,
      refreshToken: body['refreshToken'] as String?,
    );
    _currentUser = AuthonUser.fromJson(body['user'] as Map<String, dynamic>);
    return _currentUser!;
  }

  Future<void> _loadCurrentUser() async {
    final body = await _request('GET', '/v1/auth/me');
    _currentUser = AuthonUser.fromJson(body);
  }

  Future<Map<String, dynamic>> _request(
    String method,
    String path, {
    Map<String, dynamic>? body,
    bool requiresAuth = true,
  }) async {
    final uri = Uri.parse('$_baseUrl$path');
    final headers = <String, String>{
      'Content-Type': 'application/json',
      'User-Agent': 'authon-flutter/0.1.0',
    };

    if (requiresAuth && _accessToken != null) {
      headers['Authorization'] = 'Bearer $_accessToken';
    } else if (!requiresAuth) {
      headers['x-publishable-key'] = publishableKey;
    }

    late http.Response response;
    final encodedBody = body != null ? jsonEncode(body) : null;

    switch (method) {
      case 'GET':
        response = await _http.get(uri, headers: headers);
      case 'POST':
        response = await _http.post(uri, headers: headers, body: encodedBody);
      case 'PATCH':
        response = await _http.patch(uri, headers: headers, body: encodedBody);
      case 'DELETE':
        response = await _http.delete(uri, headers: headers, body: encodedBody);
      default:
        throw ArgumentError('Unsupported HTTP method: $method');
    }

    if (response.statusCode == 401 && requiresAuth && _refreshToken != null) {
      final refreshed = await _tryRefresh();
      if (refreshed) {
        headers['Authorization'] = 'Bearer $_accessToken';
        switch (method) {
          case 'GET':
            response = await _http.get(uri, headers: headers);
          case 'POST':
            response = await _http.post(uri, headers: headers, body: encodedBody);
          case 'PATCH':
            response = await _http.patch(uri, headers: headers, body: encodedBody);
          case 'DELETE':
            response = await _http.delete(uri, headers: headers, body: encodedBody);
        }
      }
    }

    if (response.statusCode >= 400) {
      Map<String, dynamic>? errorBody;
      try {
        errorBody = jsonDecode(response.body) as Map<String, dynamic>;
      } catch (_) {}
      throw AuthonException(
        statusCode: response.statusCode,
        message: errorBody?['message'] as String? ?? response.body,
        code: errorBody?['code'] as String?,
      );
    }

    if (response.body.isEmpty) return {};
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<bool> _tryRefresh() async {
    if (_refreshToken == null) return false;
    try {
      final uri = Uri.parse('$_baseUrl/v1/auth/refresh');
      final response = await _http.post(
        uri,
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'refreshToken': _refreshToken}),
      );
      if (response.statusCode == 200) {
        final body = jsonDecode(response.body) as Map<String, dynamic>;
        await _storeTokens(
          accessToken: body['accessToken'] as String,
          refreshToken: body['refreshToken'] as String?,
        );
        return true;
      }
    } catch (_) {}
    await _clearTokens();
    return false;
  }

  Future<void> _storeTokens({
    required String accessToken,
    String? refreshToken,
  }) async {
    _accessToken = accessToken;
    await _storage.write(key: _storageKeyToken, value: accessToken);
    if (refreshToken != null) {
      _refreshToken = refreshToken;
      await _storage.write(key: _storageKeyRefresh, value: refreshToken);
    }
  }

  Future<void> _clearTokens() async {
    _accessToken = null;
    _refreshToken = null;
    await _storage.delete(key: _storageKeyToken);
    await _storage.delete(key: _storageKeyRefresh);
  }

  void dispose() {
    _http.close();
  }
}

File: examples/kotlin-android/app/src/main/java/dev/authon/example/AuthonService.kt
package dev.authon.example

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.browser.customtabs.CustomTabsIntent
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit

// ---------------------------------------------------------------------------
// Data models (mobile-side; mirrors Authon REST API responses)
// ---------------------------------------------------------------------------

data class AuthUser(
    val id: String,
    val email: String? = null,
    val emailVerified: Boolean = false,
    val phone: String? = null,
    val phoneVerified: Boolean = false,
    val username: String? = null,
    val firstName: String? = null,
    val lastName: String? = null,
    val displayName: String? = null,
    val avatarUrl: String? = null,
    val banned: Boolean = false,
    val externalAccounts: List<ExternalAccount> = emptyList(),
    val createdAt: String = "",
    val updatedAt: String = "",
)

data class ExternalAccount(
    val provider: String,
    val providerId: String,
    val email: String? = null,
)

data class AuthSession(
    val id: String,
    val userId: String,
    val status: String,
    val lastActiveAt: String,
    val expireAt: String,
    val createdAt: String,
)

data class SignInRequest(
    val email: String,
    val password: String,
    val strategy: String = "password",
)

data class SignUpRequest(
    val email: String,
    val password: String,
    val firstName: String? = null,
    val lastName: String? = null,
)

data class UpdateProfileRequest(
    val firstName: String? = null,
    val lastName: String? = null,
    val username: String? = null,
)

data class MfaEnrollResponse(
    val secret: String,
    val qrCodeUrl: String,
    val backupCodes: List<String>,
)

data class AuthTokens(
    val accessToken: String,
    val refreshToken: String? = null,
    val expiresIn: Long = 3600,
)

data class ApiError(
    val message: String,
    val code: String? = null,
    @SerializedName("statusCode") val statusCode: Int = 0,
)

class AuthonException(val statusCode: Int, override val message: String, val code: String? = null) :
    Exception(message)

// ---------------------------------------------------------------------------
// Token storage
// ---------------------------------------------------------------------------

class TokenStorage(context: Context) {

    private val masterKey = MasterKey.Builder(context)
        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
        .build()

    private val prefs = EncryptedSharedPreferences.create(
        context,
        "authon_secure_prefs",
        masterKey,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )

    var accessToken: String?
        get() = prefs.getString(KEY_ACCESS_TOKEN, null)
        set(value) = prefs.edit().putString(KEY_ACCESS_TOKEN, value).apply()

    var refreshToken: String?
        get() = prefs.getString(KEY_REFRESH_TOKEN, null)
        set(value) = prefs.edit().putString(KEY_REFRESH_TOKEN, value).apply()

    fun clear() = prefs.edit().clear().apply()

    companion object {
        private const val KEY_ACCESS_TOKEN = "access_token"
        private const val KEY_REFRESH_TOKEN = "refresh_token"
    }
}

// ---------------------------------------------------------------------------
// AuthonService — HTTP client wrapping the Authon REST API
// ---------------------------------------------------------------------------

class AuthonService(context: Context) {

    private val baseUrl = BuildConfig.AUTHON_API_URL.trimEnd('/')
    private val publishableKey = BuildConfig.AUTHON_PUBLISHABLE_KEY
    val tokenStorage = TokenStorage(context)
    private val gson = Gson()
    private val applicationContext = context.applicationContext

    private val http = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .addInterceptor(HttpLoggingInterceptor().apply {
            level = if (BuildConfig.DEBUG)
                HttpLoggingInterceptor.Level.BODY
            else
                HttpLoggingInterceptor.Level.NONE
        })
        .build()

    private val json = "application/json; charset=utf-8".toMediaType()

    // -----------------------------------------------------------------------
    // Auth
    // -----------------------------------------------------------------------

    suspend fun signIn(email: String, password: String): AuthUser = withContext(Dispatchers.IO) {
        val body = gson.toJson(SignInRequest(email = email, password = password))
        val response = post("/v1/auth/sign-in", body, authenticated = false)
        val tokens = gson.fromJson(response, AuthTokens::class.java)
        tokenStorage.accessToken = tokens.accessToken
        tokenStorage.refreshToken = tokens.refreshToken
        currentUser()
    }

    suspend fun signUp(email: String, password: String, firstName: String?, lastName: String?): AuthUser =
        withContext(Dispatchers.IO) {
            val body = gson.toJson(
                SignUpRequest(
                    email = email,
                    password = password,
                    firstName = firstName?.ifBlank { null },
                    lastName = lastName?.ifBlank { null },
                )
            )
            val response = post("/v1/auth/sign-up", body, authenticated = false)
            val tokens = gson.fromJson(response, AuthTokens::class.java)
            tokenStorage.accessToken = tokens.accessToken
            tokenStorage.refreshToken = tokens.refreshToken
            currentUser()
        }

    suspend fun signOut(): Unit = withContext(Dispatchers.IO) {
        runCatching { post("/v1/auth/sign-out", "{}", authenticated = true) }
        tokenStorage.clear()
    }

    suspend fun currentUser(): AuthUser = withContext(Dispatchers.IO) {
        val response = get("/v1/auth/me")
        gson.fromJson(response, AuthUser::class.java)
    }

    suspend fun refreshSession(): Unit = withContext(Dispatchers.IO) {
        val refreshToken = tokenStorage.refreshToken
            ?: throw AuthonException(401, "No refresh token available")
        val body = gson.toJson(mapOf("refreshToken" to refreshToken))
        val response = post("/v1/auth/refresh", body, authenticated = false)
        val tokens = gson.fromJson(response, AuthTokens::class.java)
        tokenStorage.accessToken = tokens.accessToken
        tokens.refreshToken?.let { tokenStorage.refreshToken = it }
    }

    // -----------------------------------------------------------------------
    // Profile
    // -----------------------------------------------------------------------

    suspend fun updateProfile(firstName: String?, lastName: String?, username: String?): AuthUser =
        withContext(Dispatchers.IO) {
            val body = gson.toJson(
                UpdateProfileRequest(
                    firstName = firstName?.ifBlank { null },
                    lastName = lastName?.ifBlank { null },
                    username = username?.ifBlank { null },
                )
            )
            val response = patch("/v1/auth/profile", body)
            gson.fromJson(response, AuthUser::class.java)
        }

    suspend fun changePassword(currentPassword: String, newPassword: String): Unit =
        withContext(Dispatchers.IO) {
            val body = gson.toJson(
                mapOf("currentPassword" to currentPassword, "newPassword" to newPassword)
            )
            post("/v1/auth/change-password", body, authenticated = true)
            Unit
        }

    suspend fun sendPasswordResetEmail(email: String): Unit = withContext(Dispatchers.IO) {
        val body = gson.toJson(mapOf("email" to email))
        post("/v1/auth/forgot-password", body, authenticated = false)
        Unit
    }

    suspend fun deleteAccount(): Unit = withContext(Dispatchers.IO) {
        delete("/v1/auth/account")
        tokenStorage.clear()
    }

    // -----------------------------------------------------------------------
    // Sessions
    // -----------------------------------------------------------------------

    suspend fun listSessions(): List<AuthSession> = withContext(Dispatchers.IO) {
        val response = get("/v1/auth/sessions")
        val type = object : TypeToken<List<AuthSession>>() {}.type
        gson.fromJson(response, type)
    }

    suspend fun revokeSession(sessionId: String): Unit = withContext(Dispatchers.IO) {
        delete("/v1/auth/sessions/$sessionId")
    }

    // -----------------------------------------------------------------------
    // MFA
    // -----------------------------------------------------------------------

    suspend fun enrollMfa(): MfaEnrollResponse = withContext(Dispatchers.IO) {
        val response = post("/v1/auth/mfa/enroll", "{}", authenticated = true)
        gson.fromJson(response, MfaEnrollResponse::class.java)
    }

    suspend fun verifyMfaEnrollment(code: String): Unit = withContext(Dispatchers.IO) {
        val body = gson.toJson(mapOf("code" to code))
        post("/v1/auth/mfa/verify-enrollment", body, authenticated = true)
        Unit
    }

    suspend fun disableMfa(code: String): Unit = withContext(Dispatchers.IO) {
        val body = gson.toJson(mapOf("code" to code))
        post("/v1/auth/mfa/disable", body, authenticated = true)
        Unit
    }

    // -----------------------------------------------------------------------
    // OAuth / Social login
    // -----------------------------------------------------------------------

    fun launchOAuth(context: android.content.Context, provider: String) {
        val redirectUri = "${BuildConfig.AUTHON_OAUTH_REDIRECT_SCHEME}://callback"
        val url = "$baseUrl/v1/auth/oauth/$provider" +
                "?redirect_uri=${Uri.encode(redirectUri)}" +
                "&pk=${Uri.encode(publishableKey)}"

        val customTabsIntent = CustomTabsIntent.Builder()
            .setShowTitle(true)
            .build()
        customTabsIntent.launchUrl(context, Uri.parse(url))
    }

    suspend fun handleOAuthCallback(uri: Uri): AuthUser = withContext(Dispatchers.IO) {
        val code = uri.getQueryParameter("code")
            ?: throw AuthonException(400, "Missing authorization code in OAuth callback")
        val redirectUri = "${BuildConfig.AUTHON_OAUTH_REDIRECT_SCHEME}://callback"
        val body = gson.toJson(mapOf("code" to code, "redirectUri" to redirectUri))
        val response = post("/v1/auth/oauth/callback", body, authenticated = false)
        val tokens = gson.fromJson(response, AuthTokens::class.java)
        tokenStorage.accessToken = tokens.accessToken
        tokenStorage.refreshToken = tokens.refreshToken
        currentUser()
    }

    // -----------------------------------------------------------------------
    // HTTP helpers
    // -----------------------------------------------------------------------

    private fun get(path: String): String {
        val request = Request.Builder()
            .url("$baseUrl$path")
            .applyAuth()
            .get()
            .build()
        return execute(request)
    }

    private fun post(path: String, body: String, authenticated: Boolean = true): String {
        val requestBody = body.toRequestBody(json)
        val builder = Request.Builder()
            .url("$baseUrl$path")
            .post(requestBody)
        if (authenticated) builder.applyAuth()
        else builder.header("X-Authon-Publishable-Key", publishableKey)
        return execute(builder.build())
    }

    private fun patch(path: String, body: String): String {
        val requestBody = body.toRequestBody(json)
        val request = Request.Builder()
            .url("$baseUrl$path")
            .applyAuth()
            .patch(requestBody)
            .build()
        return execute(request)
    }

    private fun delete(path: String): String {
        val request = Request.Builder()
            .url("$baseUrl$path")
            .applyAuth()
            .delete()
            .build()
        return execute(request)
    }

    private fun Request.Builder.applyAuth(): Request.Builder {
        val token = tokenStorage.accessToken
        if (token != null) header("Authorization", "Bearer $token")
        return this
    }

    private fun execute(request: Request): String {
        val response = http.newCall(request).execute()
        val responseBody = response.body?.string() ?: ""
        if (!response.isSuccessful) {
            val error = runCatching { gson.fromJson(responseBody, ApiError::class.java) }.getOrNull()
            throw AuthonException(
                statusCode = response.code,
                message = error?.message ?: "HTTP ${response.code}",
                code = error?.code,
            )
        }
        return responseBody
    }
}
